test/test_clusterengine.py

```diff
@@ -0,0 +1,123 @@
+        self.assertEqual(clusterengine.cluster_count, 0)
+        self.assertEqual(clusterengine.cluster_type, ClusterType.ARTIST)
+
+    def test_get_cluster_title_negative_index(self):
+        clusterdict = ClusterDict()
+        clusterengine = ClusterEngine(clusterdict, ClusterType.ARTIST)
+        title = clusterengine.get_cluster_title(-1)
+        self.assertEqual(title, "")
+
+    def test_one_cluster(self):
+        clusterdict = ClusterDict()
+        clusterdict.add("word 1")
+        clusterdict.add("WORD  1")
+        clusterdict.add("Word 1")
+        clusterengine = ClusterEngine(clusterdict, ClusterType.ARTIST)
+        clusterengine.cluster(1.0)
+        self.assertEqual(clusterengine.cluster_count, 1)
+
+        title = clusterengine.get_cluster_title(0)
+        self.assertEqual(title, "Word 1")
+
+    def test_two_clusters(self):
+        clusterdict = ClusterDict()
+        clusterdict.add("ABC")
+        clusterdict.add("DEF")
+        clusterdict.add("abc ")
+        clusterdict.add("def")
+        clusterdict.add("abc")
+        clusterengine = ClusterEngine(clusterdict, ClusterType.ARTIST)
+        clusterengine.cluster(1.0)
+        self.assertEqual(clusterengine.cluster_count, 2)
+
+        titles = {}
+        for i in range(clusterengine.cluster_count):
+            titles[i] = clusterengine.get_cluster_title(i)
+        self.assertEqual(titles, {0: 'abc', 1: 'def'})
+
+        self.assertEqual(clusterengine.cluster_bins, {0: [0, 2, 4], 1: [1, 3]})
+        self.assertEqual(clusterengine.index_id_cluster, {2: 0, 3: 1, 0: 0, 4: 0, 1: 1})
+
+    def test_two_clusters_with_dupes(self):
+        clusterdict = ClusterDict()
+        clusterdict.add("abc")
+        clusterdict.add("def")
+        clusterdict.add("abc")
+        clusterdict.add("def")
+        clusterdict.add("abc")
+        clusterengine = ClusterEngine(clusterdict, ClusterType.ARTIST)
+        clusterengine.cluster(1.0)
+        self.assertEqual(clusterengine.cluster_count, 2)
+
+        titles = {}
+        for i in range(clusterengine.cluster_count):
+            titles[i] = clusterengine.get_cluster_title(i)
+        self.assertEqual(titles, {0: 'abc', 1: 'def'})
+
+        self.assertEqual(clusterengine.cluster_bins, {0: [0], 1: [1]})
+        self.assertEqual(clusterengine.index_id_cluster, {0: 0, 1: 1})
+
+#    def test_two_clusters_with_almost_dupes(self):
+#        clusterdict = ClusterDict()
+#        clusterdict.add("abc ")
+#        clusterdict.add("def ")
+#        clusterdict.add(" abc")
+#        clusterdict.add("def")
+#        clusterdict.add("abc")
+#
+#        clusterdict.add("abc ")
+#        clusterdict.add("def ")
+#        clusterdict.add(" abc")
+#        clusterdict.add("def")
+#        clusterdict.add("abc")
+#
+#        clusterengine = ClusterEngine(clusterdict, ClusterType.ARTIST)
+#        clusterengine.cluster(1.0)
+#        # that's a bug??
+#        self.assertEqual(clusterengine.cluster_count, 5)
```

Comment:
I think `cluster_count` is just badly named, it really should be something like `next_cluster_index`. The count of elements in `cluster_bins` should be 5 I guess, though.